text/0000-build-std/4-stage-1a.md

```diff
@@ -0,0 +1,1003 @@
+targets require `crt1.o`, `crti.o`, `crtn.o`, etc.
+
+Since [rust#76158]/[compiler-team#343], the compiler has a stable
+`-Clink-self-contained` flag which will look for special object files in
+expected locations, typically populated by the `rust-std` components. Its
+behaviour can be forced by `-Clink-self-contained=true`, but is force-enabled
+for some targets and inferred for others.
+
+Rust can start to ship `rust-self-contained` components for any targets which
+need it. These components will contain the special object files normally
+included in `rust-std`, and will be distributed for all tiers of targets. While
+generally these objects are specific to the architecture and C runtime (CRT)
+(and so `rust-self-contained-$arch-$crt` could be sufficient and result in fewer
+overall components), it's technically possible that Rust could support two
+targets with the same architecture and same CRT but different versions of the
+CRT, so having target-specific components is most future-proof. These would
+replace the `self-contained` directory in existing `rust-std` components.
+
+Similarly, for any architectures which require it, LLVM's `libunwind` will be
+built and shipped in the `rust-self-contained` component.
+
+As long as these components have been downloaded, as well as any other support
+components, such as `rust-mingw`, rustc's `-Clink-self-contained` will be able
+to link against the object files and build-std should never fail on account of
+missing special object files.
+
+*See the following sections for future possibilities:*
+
+- [*Enable local recompilation of special object files/sanitizer runtimes*][future-recompile-special]
+
+## `compiler-builtins`
+[compiler-builtins]: #compiler-builtins
+
+`compiler-builtins` is always built with `-Ccodegen-units=10000` to force each
+intrinsic into its own object file to avoid symbol clashes with libgcc. This is
+currently enforced with a profile override in the standard library's workspace.
+
+rustc will automatically use a large number of codegen units for the
+`compiler-builtins` crate, unless manually specified using the `-Ccodegen-units`
+flag (to support users, like Rust for Linux, that prefer a single codegen unit).
+This prevents `compiler-builtins` from having to be special-cased in the
+standard library workspace.
+
+> [!NOTE]
+>
+> [rust#135395] could be resurrected to implement this.
+
+### `compiler-builtins/mem`
+[compiler-builtins-mem]: #compiler-builtinsmem
+
+The `mem` feature of `compiler_builtins` (and the subsequent
+`compiler-builtins-mem` feature of `core`, `alloc`, `std` which forward to
+`compiler_builtins/mem`) is required by `no_std` crates as a `std` dependency
+will not be providing these symbols through its dependency on `libc`.
+
+It is necessary that the `compiler-builtins-mem` feature of `alloc` and/or
+`core` be enabled when `libc` is not in the crate graph
+([?][rationale-no-weak-linkage]).
```

Comment:
Managing a lot of these special cases seems concerning.  Is there some way we can simplify these down?

Reply:
In `compiler-builtins-mem`'s case, I'm not sure there's a way to avoid it, we either need `libc` to provide these or `compiler-builtins`, and if we always default to `compiler-builtins` then a locally-built std will have a noticable difference compared to a pre-built std, all else equal.

I believe it is quite common to enable `compiler-builtins-mem` as a user of `-Zbuild-std` today (e.g. see [cargo-xbuild's readme](https://github.com/rust-osdev/cargo-xbuild/blob/master/README.md#alternative-the-build-std-feature-of-cargo)) when not building `std` because you basically need to.

We explored using weak linkage to allow this feature to always be enabled but weak linkage still takes precedence over a shared object like `libc` IIRC.

Are there other special cases that you're thinking of?